File: arq_proy.cpp
#include <iostream>
#include <string>
#include <cmath>

using namespace std;

// Función para convertir un número entero a hexadecimal 
string intToHex(unsigned long long int intDecimal) {
    string hex = "";
    if (intDecimal == 0) return "0";

    char hexChars[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
    while (intDecimal > 0) {
        hex = hexChars[intDecimal % 16] + hex;
        intDecimal /= 16;
    }
    return hex;
}

// Función para convertir binario a hexadecimal 
string binaryToHex(string binary) {
    int pointPos = binary.find('.');

    string intPart = binary.substr(0, pointPos);
    string fracPart = pointPos != string::npos ? binary.substr(pointPos + 1) : "";

    // Conversión de la parte entera
    unsigned long long int intDecimal = stoull(intPart, nullptr, 2);
    string hexInteger = intToHex(intDecimal);

    // Conversión de la parte fraccionaria
    double fracDecimal = 0.0;
    if (!fracPart.empty()) {
        for (size_t i = 0; i < fracPart.size(); ++i) {
            if (fracPart[i] == '1') {
                fracDecimal += 1.0 / (1 << (i + 1));
            }
        }
    }

    string hexFractional = "";
    while (fracDecimal != 0.0) {
        fracDecimal *= 16;
        int fracInt = static_cast<int>(fracDecimal);
        hexFractional += (fracInt < 10) ? to_string(fracInt) : string(1, 'A' + fracInt - 10);
        fracDecimal -= fracInt;

        if (hexFractional.size() > 10) {  // Limitar a 10 dígitos
            break;
        }
    }

    return fracPart.empty() ? hexInteger : hexInteger + "." + hexFractional;
}

// Función para calcular el complemento restringido de un número binario
string restrictedComplement(string binary) {
    if (binary.empty()) return "Error: Cadena vacia";

    // El primer bit (bit de signo) se mantiene, los demás se invierten
    string complement = binary[0] == '0' ? "0" : "1";  // Mantiene el primer bit
    for (size_t i = 1; i < binary.size(); ++i) {
        complement += (binary[i] == '0') ? '1' : '0';  // Invertir bits
    }
    return complement;
}

// Función para convertir decimal a hexadecimal 
string decimalToHex(string decimal) {
    int pointPos = decimal.find('.');

    string intPart = decimal.substr(0, pointPos);
    string fracPart = pointPos != string::npos ? decimal.substr(pointPos + 1) : "";

    // Convertir la parte entera a hexadecimal
    unsigned long long int intDecimal = stoull(intPart);
    string hexInteger = intToHex(intDecimal);

    // Conversión de la parte fraccionaria
    double fracDecimal = fracPart.empty() ? 0.0 : stod("0." + fracPart);
    string hexFractional = "";

    for (int i = 0; fracDecimal != 0.0 && i < 20; ++i) {  // Limitar a 20 dígitos
        fracDecimal *= 16;
        int fracInt = static_cast<int>(fracDecimal);
        hexFractional += (fracInt < 10) ? to_string(fracInt) : string(1, 'A' + fracInt - 10);
        fracDecimal -= fracInt;
    }

    return fracPart.empty() ? hexInteger : hexInteger + "." + hexFractional;
}

// Función menú
void displayMenu() {
    cout << "\n===== Menu de Operaciones =====" << endl;
    cout << "1. Conversion de Binario a Hexadecimal" << endl;
    cout << "2. Obtener el Complemento Restringido de un Numero Binario" << endl;
    cout << "3. Conversion de Decimal a Hexadecimal" << endl;
    cout << "4. Salir" << endl;
    cout << "================================" << endl;
}

int main() {
    while (true) {
        displayMenu();
        int choice;
        cout << "Selecciona una opcion (1-4): ";
        cin >> choice;

        if (choice == 1) {
            string binary;
            cout << "Introduce el numero binario (ejemplo: 1010.101): ";
            cin >> binary;
            cout << "Hexadecimal: " << binaryToHex(binary) << endl;
        }
        else if (choice == 2) {
            string binary;
            cout << "Introduce el numero binario: ";
            cin >> binary;
            cout << "Complemento Restringido: " << restrictedComplement(binary) << endl;
        }
        else if (choice == 3) {
            string decimal;
            cout << "Introduce el numero decimal (ejemplo: 10.625): ";
            cin >> decimal;
            cout << "Hexadecimal: " << decimalToHex(decimal) << endl;
        }
        else if (choice == 4) {
            cout << "Gracias!" << endl;
            break;
        }
        else {
            cout << "Opcion invalida. Elige una opcion entre 1 y 4" << endl;
        }

        // Preguntar si el usuario desea realizar otra operación
        char again;
        cout << "Deseas realizar otra operacion? (s/n): ";
        cin >> again;
        if (again == 'n' || again == 'N') {
            cout << "Gracias!" << endl;
            break;
        }
    }

    return 0;
}

File: arq_proy.py
#binario a hexa (enteros y fracciones)
#complemento restringido de binario 
#decimal a hexadecimal (fracciones y periodicidad)
import sys

def binary_to_hexadecimal(binary_str):
    """
    Convierte un numero binario (con parte entera y fraccionaria) a hexadecimal.
    """
    if '.' in binary_str:
        integer_part, fractional_part = binary_str.split('.')
    else:
        integer_part, fractional_part = binary_str, ''

    # Conversión de la parte entera
    hex_integer = hex(int(integer_part, 2))[2:].upper() if integer_part else '0'

    # Conversión de la parte fraccionaria
    hex_fractional = ''
    if fractional_part:
        # Convertir la fracción binaria a decimal
        fractional_decimal = 0
        for i, digit in enumerate(fractional_part, 1):
            fractional_decimal += int(digit) * (2 ** -i)
        # Convertir la fracción decimal a hexadecimal
        fractional_decimal *= 16
        for _ in range(10):  # Limitar a 10 dígitos hexadecimales para evitar infinitas repeticiones
            digit = int(fractional_decimal)
            hex_fractional += hex(digit)[2:].upper()
            fractional_decimal = (fractional_decimal - digit) * 16
            if fractional_decimal == 0:
                break

    if hex_fractional:
        return f"{hex_integer}.{hex_fractional}"
    else:
        return hex_integer

def restricted_complement(binary_str):
    """
    Calcula el complemento restringido de un número binario.
    Suposicion: El primer bit es el bit de signo y no se invierte.
    """
    if not all(c in '01' for c in binary_str):
        return "Error: La entrada no es un numero binario valido."

    if len(binary_str) == 0:
        return "Error: Entrada vacia."

    # El primer bit (bit de signo) permanece igual
    sign_bit = binary_str[0]
    # Invertir los demás bits
    complemented_bits = ''.join('1' if bit == '0' else '0' for bit in binary_str[1:])
    return sign_bit + complemented_bits

def decimal_to_hexadecimal(decimal_str):
    """
    Convierte un numero decimal (con parte entera y fraccionaria) a hexadecimal,
    mostrando periodicidad si existe.
    """
    if '.' in decimal_str:
        integer_part, fractional_part = decimal_str.split('.')
    else:
        integer_part, fractional_part = decimal_str, ''

    # Conversión de la parte entera
    hex_integer = hex(int(integer_part))[2:].upper() if integer_part else '0'

    # Conversión de la parte fraccionaria
    if fractional_part:
        fractional_decimal = float("0." + fractional_part)
        seen = {}
        hex_fractional = ''
        repeating_start = -1

        for i in range(20):  # Limitar a 20 iteraciones para detectar repeticiones
            fractional_decimal *= 16
            digit = int(fractional_decimal)
            hex_digit = hex(digit)[2:].upper()
            hex_fractional += hex_digit
            fractional_decimal -= digit

            # Detectar periodicidad
            if fractional_decimal in seen:
                repeating_start = seen[fractional_decimal]
                break
            seen[fractional_decimal] = i + 1

            if fractional_decimal == 0:
                break

        if repeating_start != -1:
            non_repeating = hex_fractional[:repeating_start]
            repeating = hex_fractional[repeating_start:]
            hex_fractional = f"{non_repeating}({repeating})"

        return f"{hex_integer}.{hex_fractional}"
    else:
        return hex_integer

def display_menu():
    """
    Muestra el menu de opciones al usuario.
    """
    print("\n===== Menu de Operaciones =====")
    print("1. Conversion de Binario a Hexadecimal")
    print("2. Obtener el Complemento Restringido de un Numero Binario")
    print("3. Conversion de Decimal a Hexadecimal")
    print("4. Salir")
    print("================================")

def main():
    while True:
        display_menu()
        choice = input("Selecciona una opcion (1-4): ").strip()

        if choice == '1':
            binary_input = input("Introduce el numero binario (ejemplo: 1010.101): ").strip()
            result = binary_to_hexadecimal(binary_input)
            print(f"Hexadecimal: {result}")

        elif choice == '2':
            binary_input = input("Introduce el numero binario para obtener su complemento restringido: ").strip()
            result = restricted_complement(binary_input)
            print(f"Complemento Restringido: {result}")

        elif choice == '3':
            decimal_input = input("Introduce el número decimal (ejemplo: 10.625): ").strip()
            try:
                # Validar entrada decimal
                float(decimal_input)
                result = decimal_to_hexadecimal(decimal_input)
                print(f"Hexadecimal: {result}")
            except ValueError:
                print("Error: Entrada no es un numero decimal valido.")

        elif choice == '4':
            print("Gracias por usar el programa! Hasta luego.")
            sys.exit()

        else:
            print("Opcion invalida. Por favor, elige una opcion entre 1 y 4.")
            continue

        # Preguntar si el usuario desea realizar otra operación
        while True:
            again = input("Deseas realizar otra operacion? (s/n): ").strip().lower()
            if again == 's':
                break
            elif again == 'n':
                print("Gracias por usar el programa! Hasta luego.")
                sys.exit()
            else:
                print("Entrada invalida. Por favor, responde con 's' o 'n'.")

if __name__ == "__main__":
    main()
